<a href="https://colab.research.google.com/github/VitikaJain25/Eduxlabs-AI-ML/blob/main/V_eict_python_pandas_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [32]:
import pandas as pd

In [33]:
df = pd.read_csv('/content/titanic.csv')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [34]:
print(df['Age'].mean())
print(df['Age'].std())

29.69911764705882
14.526497332334042


In [35]:
print(df['Fare'].mean())
print(df['Fare'].median())
print(df['Fare'].min())
print(df['Fare'].max())

32.204207968574636
14.4542
0.0
512.3292


* Average fare paid for each category
* Average fare paid by male and female

In [36]:
list(df.groupby('Pclass')['Fare'])

[(1,
  1      71.2833
  3      53.1000
  6      51.8625
  11     26.5500
  23     35.5000
          ...   
  871    52.5542
  872     5.0000
  879    83.1583
  887    30.0000
  889    30.0000
  Name: Fare, Length: 216, dtype: float64),
 (2,
  9      30.0708
  15     16.0000
  17     13.0000
  20     26.0000
  21     13.0000
          ...   
  866    13.8583
  874    24.0000
  880    26.0000
  883    10.5000
  886    13.0000
  Name: Fare, Length: 184, dtype: float64),
 (3,
  0       7.2500
  2       7.9250
  4       8.0500
  5       8.4583
  7      21.0750
          ...   
  882    10.5167
  884     7.0500
  885    29.1250
  888    23.4500
  890     7.7500
  Name: Fare, Length: 491, dtype: float64)]

In [37]:
df.groupby('Pclass')['Fare'].mean()

Pclass
1    84.154687
2    20.662183
3    13.675550
Name: Fare, dtype: float64

In [39]:
df.groupby('Pclass')['Fare'].agg(['mean', 'median', 'min', 'max', 'std', 'var'])

mean   median  min       max        std          var
Pclass                                                           
1       84.154687  60.2875  0.0  512.3292  78.380373  6143.482816
2       20.662183  14.2500  0.0   73.5000  13.417399   180.026589
3       13.675550   8.0500  0.0   69.5500  11.778142   138.724622

In [40]:
df.groupby('Sex')['Fare'].agg(['mean', 'median', 'min', 'max', 'std', 'var'])

mean  median   min       max        std          var
Sex                                                              
female  44.479818    23.0  6.75  512.3292  57.997698  3363.732930
male    25.523893    10.5  0.00  512.3292  43.138263  1860.909702

In [41]:
df.groupby(['Pclass', 'Sex'])['Fare'].agg(['mean', 'median', 'min', 'max', 'std', 'var'])

mean    median      min       max        std          var
Pclass Sex                                                                    
1      female  106.125798  82.66455  25.9292  512.3292  74.259988  5514.545830
       male     67.226127  41.26250   0.0000  512.3292  77.548021  6013.695545
2      female   21.970121  22.00000  10.5000   65.0000  10.891796   118.631214
       male     19.741782  13.00000   0.0000   73.5000  14.922235   222.673090
3      female   16.118810  12.47500   6.7500   69.5500  11.690314   136.663446
       male     12.661633   7.92500   0.0000   69.5500  11.681696   136.462010

In [42]:
df.groupby(['Pclass', 'Sex'])['Age'].agg(['mean', 'median', 'min', 'max', 'std', 'var'])

mean  median   min   max        std         var
Pclass Sex                                                         
1      female  34.611765    35.0  2.00  63.0  13.612052  185.287955
       male    41.281386    40.0  0.92  80.0  15.139570  229.206594
2      female  28.722973    28.0  2.00  57.0  12.872702  165.706451
       male    30.740707    30.0  0.67  70.0  14.793894  218.859292
3      female  21.750000    21.5  0.75  63.0  12.729964  162.051980
       male    26.507589    25.0  0.42  74.0  12.159514  147.853777

In [43]:
df.groupby('Sex')['Age'].agg(['mean', 'median', 'min', 'max', 'std', 'var'])

mean  median   min   max        std         var
Sex                                                         
female  27.915709    27.0  0.75  63.0  14.110146  199.096233
male    30.726645    29.0  0.42  80.0  14.678201  215.449579

In [44]:
df.groupby('Pclass')['Age'].agg(['mean', 'median', 'min', 'max', 'std', 'var'])

mean  median   min   max        std         var
Pclass                                                      
1       38.233441    37.0  0.92  80.0  14.802856  219.124543
2       29.877630    29.0  0.67  70.0  14.001077  196.030152
3       25.140620    24.0  0.42  74.0  12.495398  156.134976

In [45]:
# Average of Per Passenger Fare (PPF) for each class of Passenger.
df['tota_travellers'] = df['SibSp'] + df['Parch'] + 1
df['PPF'] = df['Fare']/df['tota_travellers']

In [46]:
df.groupby('Pclass')['PPF'].mean()

Pclass
1    52.936943
2    12.698832
3     8.094756
Name: PPF, dtype: float64

**Missing Values in Data**

In [47]:
df.isnull().sum()

PassengerId          0
Survived             0
Pclass               0
Name                 0
Sex                  0
Age                177
SibSp                0
Parch                0
Ticket               0
Fare                 0
Cabin              687
Embarked             2
tota_travellers      0
PPF                  0
dtype: int64

**Drop the 'Cabin' column with 687 missing values which is close to 80% of total Data size.**

In [48]:
# drop() => used to drop any column or row where axis defines row or column.

df1 = df.drop('Cabin', axis = 'columns')
df1.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  tota_travellers       PPF  
0      0         A/5 21171   7.2500        S                2   3.62500  
1      0          PC 17599  71.2833        C                2  35.64165  
2      0  STON/O2. 3101282   7.9250        S                1   7.92500  
3      0            113803  53.1000        S                2  26.55000  
4      0            373450   8.0500        S                1   8.05000

**Drop the rows with missing values**

In [49]:
# can be used to frop rows with null values, subset represents the feature we want to consider to identify

df2 = df1.dropna(subset = ['Embarked'])
df2.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  tota_travellers       PPF  
0      0         A/5 21171   7.2500        S                2   3.62500  
1      0          PC 17599  71.2833        C                2  35.64165  
2      0  STON/O2. 3101282   7.9250        S                1   7.92500  
3      0            113803  53.1000        S                2  26.55000  
4      0            373450   8.0500        S                1   8.05000

In [50]:
df2.isnull().sum()

PassengerId          0
Survived             0
Pclass               0
Name                 0
Sex                  0
Age                177
SibSp                0
Parch                0
Ticket               0
Fare                 0
Embarked             0
tota_travellers      0
PPF                  0
dtype: int64

In [51]:
df3 = df2.copy()    #storing a copy of df2 in df3

**Fill the missing value of Age with Mean of age value**

In [52]:
round(df2['Age'].mean(),2)

29.64

In [53]:
# fillna() - to fill the null values or void spaces in any column

df2['Age'] = df2['Age'].fillna(round(df2['Age'].mean(),2))

<ipython-input-53-09aee6704865>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2['Age'] = df2['Age'].fillna(round(df2['Age'].mean(),2))


In [54]:
df2.isnull().sum()

PassengerId        0
Survived           0
Pclass             0
Name               0
Sex                0
Age                0
SibSp              0
Parch              0
Ticket             0
Fare               0
Embarked           0
tota_travellers    0
PPF                0
dtype: int64

**Filling up Missing values**
* Analyse if any categorical value affects the column with missing value. For example are age values showing some major differences with respect to the Passenger class or gender.
* Check if any outliers exist in the data

In [55]:
df3.isnull().sum()

PassengerId          0
Survived             0
Pclass               0
Name                 0
Sex                  0
Age                177
SibSp                0
Parch                0
Ticket               0
Fare                 0
Embarked             0
tota_travellers      0
PPF                  0
dtype: int64

In [60]:
df3.groupby('Sex')['Age'].agg(['mean', 'median'])

mean  median
Sex                      
female  27.745174    27.0
male    30.726645    29.0

In [61]:
df3.groupby('Pclass')['Age'].agg(['mean', 'median'])

mean  median
Pclass                   
1       38.105543    37.0
2       29.877630    29.0
3       25.140620    24.0

In [59]:
# Mean can sometimes mislead, So we use Median.

import numpy as np

n1 = np.array([1,2,2,3,4,5])
n2 = np.array([1,2,2,3,4,5,70])

print('n1-' , np.mean(n1), np.median(n1))
print('n2-' , np.mean(n2), np.median(n2))

n1- 2.8333333333333335 2.5
n2- 12.428571428571429 3.0


**Analysis made using class of pasenger represents  class 1 passengers are relatively older than passengers travelling in class 2 and class 3.**

* So we can identify or check class of passenger in which he/she is travelling to fill the missing age value according to average age of that class.

In [64]:
# Use filtering to seperate passengers with missing value of age and non missing values of age.
dfa0 = df3[df3['Age'].isnull()]
dfa1 = df3[~df3['Age'].isnull()] # Use ~ before conditon to get negation.

In [65]:
dfa0.head()

PassengerId  Survived  Pclass                           Name     Sex  Age  \
5             6         0       3               Moran, Mr. James    male  NaN   
17           18         1       2   Williams, Mr. Charles Eugene    male  NaN   
19           20         1       3        Masselmani, Mrs. Fatima  female  NaN   
26           27         0       3        Emir, Mr. Farred Chehab    male  NaN   
28           29         1       3  O'Dwyer, Miss. Ellen "Nellie"  female  NaN   

    SibSp  Parch  Ticket     Fare Embarked  tota_travellers      PPF  
5       0      0  330877   8.4583        Q                1   8.4583  
17      0      0  244373  13.0000        S                1  13.0000  
19      0      0    2649   7.2250        C                1   7.2250  
26      0      0    2631   7.2250        C                1   7.2250  
28      0      0  330959   7.8792        Q                1   7.8792

In [66]:
dfa1.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  tota_travellers       PPF  
0      0         A/5 21171   7.2500        S                2   3.62500  
1      0          PC 17599  71.2833        C                2  35.64165  
2      0  STON/O2. 3101282   7.9250        S                1   7.92500  
3      0            113803  53.1000        S                2  26.55000  
4      0            373450   8.0500        S                1   8.05000

In [69]:
def fix_missing_age_values(pclass):
  if pclass == 1:
    return 37
  elif pclass == 2:
    return 29
  else:
    return 24

In [70]:
fix_missing_age_values(3)

24

In [71]:
fix_missing_age_values(2)

29

In [73]:
# apply() => To apply a user defined function on a Data column.

dfa0['Age'] = dfa0['Pclass'].apply(fix_missing_age_values)

<ipython-input-73-4b874282ec94>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfa0['Age'] = dfa0['Pclass'].apply(fix_missing_age_values)


In [112]:
# Join back dfa0 and dfa1

df4 = pd.concat((dfa0, dfa1), axis = 'rows', ignore_index=True)
df4.head()

PassengerId  Survived  Pclass                           Name     Sex   Age  \
0            6         0       3               Moran, Mr. James    male  24.0   
1           18         1       2   Williams, Mr. Charles Eugene    male  29.0   
2           20         1       3        Masselmani, Mrs. Fatima  female  24.0   
3           27         0       3        Emir, Mr. Farred Chehab    male  24.0   
4           29         1       3  O'Dwyer, Miss. Ellen "Nellie"  female  24.0   

   SibSp  Parch  Ticket     Fare Embarked  tota_travellers      PPF  
0      0      0  330877   8.4583        Q                1   8.4583  
1      0      0  244373  13.0000        S                1  13.0000  
2      0      0    2649   7.2250        C                1   7.2250  
3      0      0    2631   7.2250        C                1   7.2250  
4      0      0  330959   7.8792        Q                1   7.8792

In [113]:
df4.isnull().sum()

PassengerId        0
Survived           0
Pclass             0
Name               0
Sex                0
Age                0
SibSp              0
Parch              0
Ticket             0
Fare               0
Embarked           0
tota_travellers    0
PPF                0
dtype: int64

**String Operations**

In [114]:
df4['Name']

0                          Moran, Mr. James
1              Williams, Mr. Charles Eugene
2                   Masselmani, Mrs. Fatima
3                   Emir, Mr. Farred Chehab
4             O'Dwyer, Miss. Ellen "Nellie"
                       ...                 
884    Rice, Mrs. William (Margaret Norton)
885                   Montvila, Rev. Juozas
886            Graham, Miss. Margaret Edith
887                   Behr, Mr. Karl Howell
888                     Dooley, Mr. Patrick
Name: Name, Length: 889, dtype: object

In [115]:
# Convert information into lower case.

df4['Name'] = df4['Name'].str.lower()

In [116]:
df4.head()

PassengerId  Survived  Pclass                           Name     Sex   Age  \
0            6         0       3               moran, mr. james    male  24.0   
1           18         1       2   williams, mr. charles eugene    male  29.0   
2           20         1       3        masselmani, mrs. fatima  female  24.0   
3           27         0       3        emir, mr. farred chehab    male  24.0   
4           29         1       3  o'dwyer, miss. ellen "nellie"  female  24.0   

   SibSp  Parch  Ticket     Fare Embarked  tota_travellers      PPF  
0      0      0  330877   8.4583        Q                1   8.4583  
1      0      0  244373  13.0000        S                1  13.0000  
2      0      0    2649   7.2250        C                1   7.2250  
3      0      0    2631   7.2250        C                1   7.2250  
4      0      0  330959   7.8792        Q                1   7.8792

In [117]:
n = 'williams, mr. charles eugene'
n.split(' ')

['williams,', 'mr.', 'charles', 'eugene']

In [118]:
n = 'williams, mr. charles eugene'
n.split(',')[0]

'williams'

In [119]:
n = 'williams, mr. charles eugene'
n.split(',')[1]

' mr. charles eugene'

In [120]:
# Create a python function which can accept the name and return surname

def get_surname(name):
  surname = name.split(',')[0]
  return surname

In [121]:
get_surname('williams, mr. charles eugene')

'williams'

In [122]:
get_surname('emir, mr. farred chehab')

'emir'

In [123]:
s = df4['Name'].apply(get_surname)
s

0           moran
1        williams
2      masselmani
3            emir
4         o'dwyer
          ...    
884          rice
885      montvila
886        graham
887          behr
888        dooley
Name: Name, Length: 889, dtype: object

In [124]:
df4.head()

PassengerId  Survived  Pclass                           Name     Sex   Age  \
0            6         0       3               moran, mr. james    male  24.0   
1           18         1       2   williams, mr. charles eugene    male  29.0   
2           20         1       3        masselmani, mrs. fatima  female  24.0   
3           27         0       3        emir, mr. farred chehab    male  24.0   
4           29         1       3  o'dwyer, miss. ellen "nellie"  female  24.0   

   SibSp  Parch  Ticket     Fare Embarked  tota_travellers      PPF  
0      0      0  330877   8.4583        Q                1   8.4583  
1      0      0  244373  13.0000        S                1  13.0000  
2      0      0    2649   7.2250        C                1   7.2250  
3      0      0    2631   7.2250        C                1   7.2250  
4      0      0  330959   7.8792        Q                1   7.8792

In [125]:
# insert the value in s after Name column
df4.insert(4, 'Surname', s)

In [108]:
df4.head()

PassengerId  Survived  Pclass                           Name    Surname  \
0            6         0       3               moran, mr. james     Braund   
1           18         1       2   williams, mr. charles eugene    Cumings   
2           20         1       3        masselmani, mrs. fatima  Heikkinen   
3           27         0       3        emir, mr. farred chehab   Futrelle   
4           29         1       3  o'dwyer, miss. ellen "nellie"      Allen   

      Sex   Age  SibSp  Parch  Ticket     Fare Embarked  tota_travellers  \
0    male  24.0      0      0  330877   8.4583        Q                1   
1    male  29.0      0      0  244373  13.0000        S                1   
2  female  24.0      0      0    2649   7.2250        C                1   
3    male  24.0      0      0    2631   7.2250        C                1   
4  female  24.0      0      0  330959   7.8792        Q                1   

       PPF  
0   8.4583  
1  13.0000  
2   7.2250  
3   7.2250  
4   7.8792